controllers/drplacementcontrol_controller.go

```diff
@@ -2422,3 +2447,181 @@ func (r *DRPlacementControlReconciler) determineDRPCState(
+	cluster string,
+	drpc *rmn.DRPlacementControl,
+	vrgNamespace string,
+) {
+	log.Info("adopting existing VRG ManifestWork", "cluster", cluster, "namespace", vrg.Namespace, "name", vrg.Name)
+
+	if vrg.GetAnnotations() == nil {
+		vrg.Annotations = make(map[string]string)
+	}
+
+	if v, ok := vrg.Annotations[DRPCUIDAnnotation]; ok && v == string(drpc.UID) {
+		// Annotation may already be set but not reflected on the resource view yet
+		log.Info("detected VRGs DRPC UID annotation as existing",
+			"cluster", cluster, "namespace", vrg.Namespace, "name", vrg.Name)
+
+		return
+	}
+
+	vrg.Annotations[DRPCUIDAnnotation] = string(drpc.UID)
+
+	annotations := make(map[string]string)
+	annotations[DRPCNameAnnotation] = drpc.Name
+	annotations[DRPCNamespaceAnnotation] = drpc.Namespace
+
+	err := mwu.CreateOrUpdateVRGManifestWork(drpc.Name, vrgNamespace, cluster, *vrg, annotations)
+	if err != nil {
+		log.Info("error updating VRG via ManifestWork during adoption", "error", err, "cluster", cluster)
+	}
+}
+
+// adoptOpphanVRG creates a missing ManifestWork for a VRG found via a ManagedClusterView
+func adoptOrphanVRG(
+	log logr.Logger,
+	mwu rmnutil.MWUtil,
+	viewVRG *rmn.VolumeReplicationGroup,
+	cluster string,
+	drpc *rmn.DRPlacementControl,
+	vrgNamespace string,
+) {
+	log.Info("adopting orphaned VRG ManifestWork",
+		"cluster", cluster, "namespace", viewVRG.Namespace, "name", viewVRG.Name)
+
+	annotations := make(map[string]string)
+	annotations[DRPCNameAnnotation] = drpc.Name
+	annotations[DRPCNamespaceAnnotation] = drpc.Namespace
+
+	// Adopt the namespace as well
+	err := mwu.CreateOrUpdateNamespaceManifest(drpc.Name, vrgNamespace, cluster, annotations)
+	if err != nil {
+		log.Info("error creating namespace via ManifestWork during adoption", "error", err, "cluster", cluster)
+
+		return
+	}
+
+	vrg := constructVRGFromView(viewVRG)
+	if vrg.GetAnnotations() == nil {
+		vrg.Annotations = make(map[string]string)
+	}
+
+	vrg.Annotations[DRPCUIDAnnotation] = string(drpc.UID)
+
+	if err := mwu.CreateOrUpdateVRGManifestWork(
+		drpc.Name, vrgNamespace,
+		cluster, *vrg, annotations); err != nil {
+		log.Info("error creating VRG via ManifestWork during adoption", "error", err, "cluster", cluster)
+	}
+}
+
+// constructVRGFromView selectively constructs a VRG from a view, using its spec and only those annotations that
+// would be set by the hub on the ManifestWork
+func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplicationGroup {
+	vrg := &rmn.VolumeReplicationGroup{
+		TypeMeta: metav1.TypeMeta{Kind: "VolumeReplicationGroup", APIVersion: "ramendr.openshift.io/v1alpha1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      viewVRG.Name,
+			Namespace: viewVRG.Namespace,
+		},
+	}
+
+	viewVRG.Spec.DeepCopyInto(&vrg.Spec)
+
+	for k, v := range viewVRG.GetAnnotations() {
+		switch k {
+		case DestinationClusterAnnotationKey:
+		case DoNotDeletePVCAnnotation:
+		case DRPCUIDAnnotation:
+			rmnutil.AddAnnotation(vrg, k, v)
```

Comment:
You need to save all 3 annotations right? So don't you need the keyword `fallthrough` for case 1 and 2?

Reply:
done